README.md
Name
====

OpenWAF

The first all-round open source Web security protection system, more protection than others.

Table of Contents
=================

* [Name](#name)
* [Version](#version)
* [Synopsis](#synopsis)
* [Description](#description)
* [Installation](#installation)
* [Community](#community)
    * [English Mailing List](#english-mailing-list)
    * [Chinese Mailing List](#chinese-mailing-list)
    * [Personal QQ Mail](#personal-qq-mail)
* [Bugs and Patches](#bugs-and-patches)
* [TODO](#todo)
* [Changes](#changes)
* [Copyright and License](#copyright-and-license)
* [Modules Configuration Directives](#modules-configuration-directives)
* [Nginx Variables](#nginx-variables)
* [SecRules](#secrules)
    * [Variables](#variables)
    * [Transformation Functions](#transformation-functions)
    * [Operators](#operators)
    * [Others](#others)
* [Donation](#donation)

Version
=======

This document describes OpenWAF v1.0.0β released on 21 Feb 2020.

Docker Version  
1. titansec/openwaf:0.0.1.161130_beta  
&emsp;&emsp;SHA: 596dee9d2b9ce44d59dc445141f72b3607f9fbe6  
&emsp;&emsp;https://github.com/titansec/OpenWAF/tree/596dee9d2b9ce44d59dc445141f72b3607f9fbe6  
2. titansec/openwaf:0.0.3.170103_beta  
&emsp;&emsp;SHA: 28ce1556250301f26f31b46d9cd9dde5a3b3f03f  
&emsp;&emsp;https://github.com/titansec/OpenWAF/tree/28ce1556250301f26f31b46d9cd9dde5a3b3f03f  
3. titansec/openwaf:v0.0.4  
&emsp;&emsp;SHA: cdef1e0eac4f12c374feb6ac3c3c465610daaf6c  
&emsp;&emsp;https://github.com/titansec/OpenWAF/tree/cdef1e0eac4f12c374feb6ac3c3c465610daaf6c  
4. titansec/openwaf:v0.0.5  
&emsp;&emsp;https://github.com/titansec/OpenWAF/tree/f9a3a6cfb8b75b1f4af9f44477a5b64337a7da47  
5. titansec/openwaf:v0.0.6(titansec/openwaf:latest)  
&emsp;&emsp;https://github.com/titansec/OpenWAF/tree/7c86321c69e2423cacc432e7f41cf37ec4b42c6b  

Synopsis
========
```nginx
    #nginx.conf
    lua_package_path '/twaf/?.lua;;';
    
    init_by_lua_file /twaf/app/twaf_init.lua;
    
    lua_shared_dict twaf_shm 50m;
    
    upstream test {
        server 0.0.0.1; #just an invalid address as a place holder
        balancer_by_lua_file twaf_balancer.lua;
    }
    
    server {
        listen 443 ssl;
        server_name _;
        
        ssl_certificate_by_lua_file  twaf_ssl_cert.lua;
        rewrite_by_lua_file          /twaf/app/twaf_rewrite.lua;
        access_by_lua_file           /twaf/app/twaf_access.lua;
        header_filter_by_lua_file    /twaf/app/twaf_header_filter.lua;
        body_filter_by_lua_file      /twaf/app/twaf_body_filter.lua
        log_by_lua_file              /twaf/app/twaf_log.lua;
        
        set $twaf_https 1;
        set $twaf_upstream_server "";
        
        ssl_certificate nginx.crt;
        ssl_certificate_key nginx.key;
        
        location / {
            lua_need_request_body on;
            proxy_pass $twaf_upstream_server;
        }
    }
    
    server {
        listen      80;
        server_name  _;
        
        rewrite_by_lua_file       /twaf/app/twaf_rewrite.lua;
        access_by_lua_file        /twaf/app/twaf_access.lua;
        header_filter_by_lua_file /twaf/app/twaf_header_filter.lua;
        body_filter_by_lua_file   /twaf/app/twaf_body_filter.lua
        log_by_lua_file           /twaf/app/twaf_log.lua;
        
        set $twaf_upstream_server "";
        
        location / {
            lua_need_request_body on;
            proxy_pass $twaf_upstream_server;
        }
    }
```

```json
    #default_config-json

    #main_safe_policy-json
```

[Back to TOC](#table-of-contents)

Description
===========

&emsp;&emsp;OpenWAF is the first fully open source Web application protection system (WAF), based on nginx_lua API analysis of HTTP request information. OpenWAF is composed of two functional engines: behavior analysis engine and rule engine. The rule engine mainly analyzes the individual requests, and the behavior analysis engine is mainly responsible for the tracking of the request information.  
&emsp;&emsp;Rule engine inspired by [modsecurity](https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual) and [freewaf(lua-resty-waf)](https://github.com/p0pr0ck5/lua-resty-waf), the ModSecurity rules will be implemented using lua. The rule engine can be based on the protocol specification, automatic tools, injection attacks, cross site attacks, information leaks and other security exception request, adding support for dynamic rules, timely repair vulnerabilities.  
&emsp;&emsp;Behavior analysis engine including fuzzy identification based on frequency, anti malware crawler, human-computer identification anti detection module, anti CSRF, anti CC, anti right, protection against attack file upload module, cookie tamper proof, anti-theft chain, custom headers and attack response page proof module of information disclosure.  
&emsp;&emsp;In addition to the two engines, but also includes statistics, log, attack response page, access rules and other basic modules. In addition to the existing functional modules, OpenWAF also supports dynamic modification of the configuration, the dynamic addition of third party modules, so that the engine does not restart under the conditions of the outage, upgrade protection.  
&emsp;&emsp;OpenWAF supports the above features as a strategy for different web application applications with different strategies to protect. The future will build a cloud platform, the strategy can also be shared for others.

basic modules:
* [openwaf_conf](https://github.com/titansec/openwaf_conf)
* [openwaf_log](https://github.com/titansec/openwaf_log)
* [openwaf_reqstat](https://github.com/titansec/openwaf_reqstat)
* [openwaf_core](https://github.com/titansec/openwaf_core)
* [openwaf_access_rule](https://github.com/titansec/openwaf_access_rule)

safe modules:
* [openwaf_rule_engine](https://github.com/titansec/openwaf_rule_engine)
* [openwaf_attack_response](https://github.com/titansec/openwaf_attack_response)
* [openwaf_api](https://github.com/titansec/openwaf_api)
* [openwaf_anti_mal_crawler](https://github.com/titansec/openwaf_anti_mal_crawler)
* [openwaf_anti_cc](https://github.com/titansec/openwaf_anti_cc)

Detailed configuration documents and examples, please refer to the above module documentation
  
[Back to TOC](#table-of-contents)

Installation
============

[请看 OpenWAF 安装文档](https://github.com/titansec/OpenWAF/blob/master/doc/%E8%BD%BB%E6%9D%BE%E7%8E%A9%E8%BD%ACOpenWAF%E4%B9%8B%E5%AE%89%E8%A3%85%E7%AF%87.md)

[Back to TOC](#table-of-contents)

Community
=========

English Mailing List
--------------------

The [OpenWAF-en](https://groups.google.com/group/openwaf-en) mailing list is for English speakers.

Chinese Mailing List
--------------------

The [OpenWAF-cn](https://groups.google.com/group/openwaf-cn) mailing list is for Chinese speakers.

Personal QQ Mail
----------------

290557551@qq.com

QQ Group
---------

579790127

[Back to TOC](#table-of-contents)

Bugs and Patches
================

Please submit bug reports, wishlists, or patches by

1. creating a ticket on the [GitHub Issue Tracker](https://github.com/titansec/OpenWAF/issues),
1. or posting to the [OpenWAF community](#community).

[Back to TOC](#table-of-contents)

TODO
====

* Update OpenResty version to 1.15.8.2.Time: February 24, 2020 to February 28, 2020
* Update docker version to v1.0.0 β. Time: February 24, 2020 to February 28, 2020
* Support connecting SSO
* Dynamic token
* APISG(API Security gateway)
* Mock

[Back to TOC](#table-of-contents)

Changes
=======

[Changelog](https://github.com/titansec/OpenWAF/blob/master/Changelog)

[Back to TOC](#table-of-contents)

Copyright and License
=====================

This module is licensed under the BSD license.

Copyright (C) 2016-2016, by Jian "Miracle" Qi (齐健) <miracleqi25@gmail.com>, Titan Co.Ltd.

All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

[Back to TOC](#table-of-contents)

Modules Configuration Directives
================================
* [twaf_access_rule](#twaf_access_rule)
* [twaf_anti_hotlink](#twaf_anti_hotlink)
* [twaf_anti_mal_crawler](#twaf_anti_mal_crawler)
* [twaf_reqstat](#twaf_reqstat)
* [twaf_log](#twaf_log)
* [twaf_secrules](#twaf_secrules)
* [twaf_anti_cc](#twaf_anti_cc)

[Back to TOC](#table-of-contents)

twaf_access_rule
----------------

```txt
{
    "twaf_access_rule": {
        "rules": [                                 -- 注意先后顺序
            {                                      
                "user": "user_id",                 -- 用户名ID，非必填，默认值"-"
                "ngx_ssl": false,                  -- nginx认证的开关，非必填，默认值false
                "ngx_ssl_cert": "path",            -- nginx认证所需PEM证书地址
                "ngx_ssl_key": "path",             -- nginx认证所需PEM私钥地址
                "host": "^1\\.1\\.1\\.1$",         -- 域名，支持正则匹配，支持字符串或数组，同时支持IPv4/IPv6
                "port": 80,                        -- 端口号。支持number或数组类型，非必填，默认值80或443
                "path": "\/",                      -- 路径，支持正则匹配，非必填，默认值"/"
                "url_case_sensitive": false,       -- 路径区分大小写，boolean类型，非必填，默认值 false
                "server_ssl": false,               -- 后端服务器ssl开关，boolean类型，非必填，默认值 false
                "forward": "server_5",             -- 后端服务器upstream名称，string类型
                "forward_addr": "1.1.1.2",         -- 后端服务器ip地址，string类型
                "forward_port": "8080",            -- 后端服务器端口号，非必填，默认值80或443
                "uuid": "access_567b067ff2060",    -- 用来标记此规则的uuid，非必填，默认16位随机字符串
                "policy": "policy_uuid"            -- 安全策略ID，string类型，非必填，默认值twaf_default_conf
            }
        ]
    }
}
```
rules
-----
**syntax:** *"rules": table*

**default:** *none*

**context:** *twaf_access_rule*

table类型，接入规则，顺序匹配

user
----
**syntax:** *"user": string*

**default:** *-*

**context:** *twaf_access_rule:rules*

string类型。用户名ID。对应变量 %{USERID}。

非必填，默认值"-".

ngx_ssl
-------
**syntax:** *"ngx_ssl": true|false*

**default:** *false*

**context:** *twaf_access_rule:rules*

boolean类型，服务器端(nginx)认证开关，可与client_ssl组成双向认证

非必填，默认值false

ngx_ssl_cert
------------
**syntax:** *"ngx_ssl_cert": "path"*

**default:** *none*

**context:** *twaf_access_rule:rules*

string类型，服务器端(nginx)认证所需PEM证书地址，目前仅支持绝对地址

ngx_ssl_key
-----------
**syntax:** *"ngx_ssl_key": "path"*

**default:** *none*

**context:** *twaf_access_rule:rules*

string类型，服务器端(nginx)认证所需PEM私钥地址，目前仅支持绝对地址

host
----
**syntax:** *"host": "ip|domain name regex"*

**default:** *none*

**context:** *twaf_access_rule:rules*

string或数组类型。发布域名。(从 v1.0.0β 版本开始支持数组)

支持正则表达式(匹配时，自动忽略大小写)。

同时支持IPv4/IPv6。(从 v1.0.0β 版本开始支持IPv6)

例如:
```
    "host": "^1\\.1\\.1\\.1$"
    "host": "test\\.com"
    "host": "^.*\\.com$"
    "host": "www.baidu.com"
    "host": ["www.baidu.com", "1.1.1.1", "8888::192.168.1.1"]
```

port
----
**syntax:** *"port": number*

**default:** *80|443*

**context:** *twaf_access_rule:rules*

number或数组类型，端口号。(从 v1.0.0β 版本开始支持数组)

非必填，默认值80或443

若"host"参数为数组时，"port"也应为数组。例如:
```
    "host": ["www.baidu.com", "1.1.1.1", "8888::192.168.1.1"]
    "posrt": [80, 8088, 8099]
```

path
----
**syntax:** *"path": "regex"*

**default:** *none*

**context:** *twaf_access_rule:rules*

string类型，路径，支持正则匹配

非必填，默认值"/"

例如:
```
    "path": "/"
    "path": "/images"
    "path": "/[a|b]test"
```

url_case_sensitive
------------------
**syntax:** *"url_case_sensitive": "true|false"*

**default:** *false*

**context:** *twaf_access_rule:rules*

boolean类型，路径区分大小写(从 v1.0.0β 版本开始支持此参数)

非必填，默认值false(不区分大小写)

server_ssl
----------
**syntax:** *"server_ssl": true|false*

**default:** *false*

**context:** *twaf_access_rule:rules*

boolean类型，OpenWAF向后端服务器连接的ssl开关

非必填，默认值 false

例如:
```
    upstream test {
    	server 1.1.1.1;
    }
    
    http {
    	server {
    	    listen 80;
    	    server_name _;
    	    
    	    location / {
    	        #server_ssl为true，相当于proxy_pass后为https
    	    	proxy_pass https://test;
    	        #server_ssl为false，相当于proxy_pass后为http
    	    	#proxy_pass http://test;
    	    }
    	}
    }
```

forward
-------
**syntax:** *"forward": "upstream_uuid"*

**default:** *none*

**context:** *twaf_access_rule:rules*

string类型，forward表示后端服务器的uuid，即upstream的名称

若不使用OpenWAF提供的$twaf_upstream_server变量，则"forward","forward_addr","forward_port"均非必填(从 v1.0.0β 版本开始支持非必填)

```
    #如：使用OpenWAF自带的$twaf_upstream_server变量，forward值为test
    upstream test {
        server 1.1.1.1;
    }
    
    server {
        ...
        location / {
            proxy_pass $twaf_upstream_server;
        }
    }
    
    ---------------------------------
    
    #如：未使用OpenWAF自带的$twaf_upstream_server变量，forward非必填
    server {
        ...
        location / {
            proxy_pass http://1.1.1.1;
        }
    }
    
    ---------------------------------
    
    #如：未使用OpenWAF自带的$twaf_upstream_server变量，forward非必填
    server {
        ...
        location / {
            root html;
            index index.htm;
        }
    }
```

forward_addr
------------
**syntax:** *"forward_addr": "ip"*

**default:** *none*

**context:** *twaf_access_rule:rules*

string类型，forward_addr表示后端服务器的ip地址（TODO：支持域名）

```
    upstream test {
        #如：forward_addr值为1.1.1.1
    	server 1.1.1.1;
    }
```

forward_port
------------
**syntax:** *"forward_port": port*

**default:** *80|443*

**context:** *twaf_access_rule:rules*

number类型，forward_port表示后端服务器端口号

非必填，默认值80或443

```
    upstream test {
    	#如：forward_port值为50001
    	server 1.1.1.1:50001;
    }
```

uuid
----
**syntax:** *"uuid": "string"*

**default:** *random(16)*

**context:** *twaf_access_rule:rules*

string类型，接入规则的唯一标识

非必填，默认16位随机字符串(从 v1.0.0β 版本开始默认16位随机字符串)

policy
------
**syntax:** *"policy": "policy_uuid"*

**default:** *twaf_default_conf*

**context:** *twaf_access_rule:rules*

string类型，满足此接入规则的请求，所使用安全策略的uuid

非必填，默认值twaf_default_conf

[Back to twaf_access_rule](#twaf_access_rule)

[Back to TOC](#table-of-contents)

twaf_anti_hotlink
-----------------
```json
{
    "twaf_anti_hotlink":{
        "state":false,
        "log_state":true,
        "event_id":"110001",
        "event_severity":"medium",
        "ct_state":false,
        "action_meta":403,
        "action":"DENY",
        "mode":"referer",
        "allow_noreferer":true,
        "cookie_name":"TWAF_AH",
        "uri_ext":["javascript", "css", "html", ""]
    }
}
```
state
-----
**syntax:** *"state": true|false|"$dynamic_state"*

**default:** *false*

**context:** *twaf_anti_hotlink*

当前模块暂未开源

log_state
---------
**syntax:** *"log_state": true|false|"$dynamic_state"*

**default:** *true*

**context:** *twaf_anti_hotlink*

ct_state
--------
**syntax:** *"ct_state": true|false|"$dynamic_state"*

**default:** *false*

**context:** *twaf_anti_hotlink*

event_id
--------
**syntax:** *"event_id": "string"*

**default:** *"110001"*

**context:** *twaf_anti_hotlink*

event_severity
--------------
**syntax:** *"event_severity": "string"*

**default:** *"medium"*

**context:** *twaf_anti_hotlink*

action
------
**syntax:** *"action": "string"*

**default:** *"DENY"*

**context:** *twaf_anti_hotlink*

action_meta
-----------
**syntax:** *"action_meta": "string"|number*

**default:** *403*

**context:** *twaf_anti_hotlink*

mode
----
**syntax:** *"mode": "string"*

**default:** *"referer"*

**context:** *twaf_anti_hotlink*

allow_noreferer
---------------
**syntax:** *"allow_noreferer": true|false*

**default:** *true*

**context:** *twaf_anti_hotlink*

cookie_name
-----------
**syntax:** *"cookie_name": "string"*

**default:** *TWAF_AH*

**context:** *twaf_anti_hotlink*

cookie_name表示盗链模块发送COOKIE的名称，默认"TWAF_AH"

此配置只有mode为cookie模式下生效

uri_ext
-------
**syntax:** *"uri_ext": array|exten|"all"*

**default:** *none*

**context:** *twaf_anti_hotlink*

uri_ext表示对哪些资源进行盗链防护

```
     #对html类型资源进行盗链防护
     "uri_ext": "html"
     
     #对未知类型资源进行盗链防护，nginx无法解析出资源类型时为空字符串
     "uri_ext": ""
     
     #对html、css及未知类型资源进行盗链防护
     "uri_ext": ["html", "css", ""]
     
     #对所有资源进行盗链防护
     "uri_ext": "all"
```

[Back to twaf_anti_hotlink](#twaf_anti_hotlink)

[Back to TOC](#table-of-contents)

twaf_anti_mal_crawler
---------------------
```txt
{
    "state": false,                                            -- 模块开关，支持 true，false
    "log_state":true,                                          -- 日志开关

    "dict_state": false,                                       -- shared_dict 开关
    "shared_dict_name":"twaf_anti_mal_crawler",                -- shared_dict 名称,若为空，则值为 "twaf_global" 下的 "dict_name"
    "shared_dict_key": "remote_addr",                          -- shared_dict 键值
    "timeout":300,                                             -- shared_dict 保存状态有效时长（单位秒）
    "timer_flush_expired":200,                                 -- shared_dict 清除过期信息的间隔时间（单位秒）,若为空，则值为 "twaf_global" 下的 "timer_flush_expired"

    "cookie_state":true,                                       -- cookie机制开关
    "crawler_cookie_name":"TWAF_crawler",                      -- 爬虫cookie名称
    "mal_cookie_name":"TWAF_mcrawler",                         -- 恶意爬虫cookie名称


Changelog
Time: 2020/02/14
Version: v1.0.0β
文档已更新：
    twaf_access_rule 接入规则模块：
        Add: 新增 url_case_sensitive 参数，用于判断路径 "path" 是否区分大小写(默认不区分大小写)
        Add: host 参数与 port 参数支持数组类型
        Add: host 参数支持 IPv6
        Add: 接入规则 POST 和 DELETE 的 API 支持批量添加或批量删除接入规则
        Update: 接入规则 PUT API由 部分修改配置 改为 全量修改配置
        Add: 新增接入规则的 PATCH API 用于部分修改配置

    twaf_secrules 规则引擎模块：
        Add: 规则引擎中 opts 的 ngx_var 支持动态变量
        Add: 规则引擎中 operators 的 ip_utils 支持 IPv6
        Add: 添加规则集机制。可设置不同规则集，并在策略中引用指定规则集。(策略未引用规则集，则默认加载所有规则生效，兼容了无规则集的旧版本)
        Add: 新增规则集APIs。包含增删改查。

    twaf_log 日志模块：
        Add: 日志支持写入本地文件(同时支持原有tcp/udp外发)

    twaf_stat 统计模块：
        Update: 统计：支持按关键字进行分类统计。

    twaf_api API模块：
        Add: 新增策略policy的PATCH方法API，用于修改策略部分配置。
        Add: 新增 errlog API，可通过调用API查询错误日志
        Add: 新增luajit API，便于判断环境是 lua 还是 luajit
        Add: 新增PSET功能，可自定义“对象”
        Update: user_defined_rules 的 POST API 不再支持 {index} 参数

    others：
        Update: 变量 UNIQUE_ID，默认34位自定义随机字符串 改为从 $request_id 变量获取的 16/32 位随机字符串
        Add: 默认添加了 X-Tt-Request-Id 响应头，其值等于 UNIQUE_ID 变量的值.
        Add: 可从XFF中获取真实来源IP。开启接入规则模块，此功能才生效。 ?????? twaf_global 在哪描述？
        Fix: [warn] could not build optimal variables_hash。在nginx配置文件中放大variables_hash_max_size和variables_hash_bucket_size的值，解决此问题。

文档未更新：
    Add: 支持响应体内容替换
    Update: OpenWAF 颁发的cookie，默认追加HttpOnly属性。
    Add: 可用规则生成不同频率级别的CC防护
    Update: 自定义响应头支持配置动态变量
    Update:“重定向”动作支持配置动态变量
    Update: 变量，按需获取。不必每个请求都加载所有变量
    Add: 配置初始化。无需在请求过程中处理不必要的配置校验及转换。
    Add: 基于配置初始化，规则引擎设置二级缓存(变量缓存级transform结果缓存)
    Add: 基于配置初始化，接入规则设置二级缓存，实现快速匹配(由原来的顺序匹配O(N)，改为缓存定位O(1))


Time: 2017/12/26  
Version: v0.0.6  
    Access_rule module :
        fix : 'host' not ignore case
    
    Log module :
        add directive : 'size_limit'
    
    Api module :
        upload API module
        fix : failed to request api
    
    Anti_mal_crawler module :
        fix : mistakenly identified as crawler
        add directive : 'dict_state'
    
    Anti_cc module :
        fix : no log in CC protection period
        add directive : 'attacks' 
    
    Secrules module :
        add action : 'WARN'
        add action : 'AUDIT'
        add action : 'ALLOW_PHASE'
        add directive : 'system_rules_state'
        add directive : 'recommend' in rules
        add directive : 'add_resp_headers' in rules
        fix : 'meta' not work when action is 'DENY'
        change the order of system_rule and user_defined_rules
    
    Others :
       upload doc : 轻松玩转OpenWAF之ELK
       upload doc : 深入研究OpenWAF之nginx配置
       upload doc : 轻松玩转OpenWAF之安装篇
       upload donation
       support waf bypass
       fix : return 500, ngx.req.raw_header() not support HTTP/2
       create CODE_OF_CONDUCT.md
       create CONTRIBUTING.md
       update docker version to 0.0.6
    
    New Release : 0.0.6
    
Time: 2017/04/10  
Version: v0.0.5  
    1. Rules
        Delete 30 rules which severity is "low".
    2. fix
        Failed to install on Ubuntu: undefined symbol GeoIP_continent_by_id
    3. Update docker version to 0.0.5
    4. New Release - 0.0.5
    
Time: 2017/03/20  
Version: v0.0.4  
    1. New Module - twaf_anti_cc  
        Anti http flood  
    2. Performance optimization
        optimaize RESPONSE_BODY variable to reduce memory
    3. add QQ group
        QQ group: 579790127
    4. fix
        4.1 loading error -- wrong comment
        4.2 return 500 response code -- string.char not support GBK
    5. Update docker version
    6. New Release - 0.0.4
    
Time: 2017/01/03  
Version: v0.0.3.170103_beta  
    1. New Module - twaf_anti_mal_crawler  
        Distinguish malicious crawler and some scan tools 
    
Time: 2016/12/05  
Version: v0.0.2.161205_beta  
    1. New Module - twaf_attack_response  
        Return custom response page When the request is rejected by OpenWAF  
    2. Api - api/stat[/policy_uuid]  
        Show statistical infomation  
    
Time: 2016/12/05  
Version: v0.0.1.161130_beta  
    1. Docker  
        build OpenWAF with docker  
    
Time: 2016/12/05  
Version: v0.0.1.161012_beta  
    1. log module  
        Send tcp/udp log  
    2. reqstat module  
        Statistics of request infomation  
    3. access rule  
        Publish applications  
    4. rule engine  
        Access Control  
    

lib/twaf/twaf_core.lua

-- Copyright (C) Miracle
-- Copyright (C) OpenWAF

local _M = {
    _VERSION = "1.1.0"
}

local cjson                                = require "cjson"
local twaf_func                            = require "lib.twaf.inc.twaf_func"
local twaf_request                         = require "lib.twaf.inc.request"
local twaf_opts                            = require "lib.twaf.inc.opts"
local lrucache                             = require "resty.lrucache"

local mt            = { __index = _M, }
local ngx_log       = ngx.log
local ngx_WARN      = ngx.WARN
local ngx_ERR       = ngx.ERR
local ngx_DONE      = ngx.DONE
local ngx_time      = ngx.time
local ngx_get_phase = ngx.get_phase
local ngx_header    = ngx.header
local ngx_ERROR     = ngx.ERROR
local ngx_var       = ngx.var
local string_format = string.format
local table_insert  = table.insert
local table_concat  = table.concat
local _tostring     = tostring
local _type         = type

function _M.new(self, config)

    self.modfactory            = {}
    self.access_modules        = {}
    self.header_filter_modules = {}
    
    local lruc, err = lrucache.new(200)
    if not lruc then
        return error("failed to create the cache: " .. (err or "unknown"))
    end

    return setmetatable({
        config = config,
        lruc   = lruc
    }, mt)
end

-- A safe place in ngx.ctx for the current module instance (self).
function _M.ctx(self)
    local ctx = ngx.ctx
    if not ctx.req then
        ctx.req = {
            TX = {},
            MATCHED_VARS = {},
            MATCHED_VAR_NAMES = {},
            POLICYID = self.config.global_conf_uuid,
            NGX_VAR = ngx_var
        }
        ctx.storage = {}
        ctx.debug   = {}
        ctx.events  = {
            log     = {},
            stat    = {},
            info    = ""
        }
    end
    return ctx
end

function _M.get_vars(self, var, req)
    return req[var] or (twaf_request.vars[var] and twaf_request.vars[var](req) or nil)
end

function _M.get_vars_no_cache(self, var, req)
    return twaf_request.vars[var] and twaf_request.vars[var](req) or nil
end

function _M.register_modules(self, modules)

    if _type(modules) ~= "table" then
        return false
    end
    
    for modules_name, path in pairs(modules) do
        self.modfactory[modules_name] = require(path)
    end
    
    return true
end

function _M.get_default_config_param(self, param)
    return self.config.twaf_default_conf[param]
end

function _M.get_config_param(self, param)

    local policy_uuid = self:ctx().req.POLICYID
    
    if policy_uuid and policy_uuid ~= "twaf_default_conf" then
        local policy = self.config.twaf_policy[policy_uuid]
        if policy == nil then
            ngx_log(ngx_ERR, "No policy: "..policy_uuid)
            return
        end
        
        return policy[param]
    end
    
    return self.config.twaf_default_conf[param]
end

function _M.get_modules_config_param(self, modules, param)

    local policy_uuid = self:ctx().req.POLICYID
    
    if policy_uuid and policy_uuid ~= "twaf_default_conf" then
        local policy = self.config.twaf_policy[policy_uuid]
        if policy == nil then
            ngx_log(ngx_ERR, "No policy: "..policy_uuid)
            ngx.exit(500)
            return
        end
        
        local mod = policy[modules]
        if mod == nil then
            ngx_log(ngx_ERR, string_format("No module '%s' in policy '%s'", modules, policy_uuid))
            ngx.exit(500)
            return
        end
        
        return mod[param]
    end
    
    local mod = self.config.twaf_default_conf[modules]
    if mod == nil then
        ngx_log(ngx_ERR, string_format("No module '%s' in policy 'twaf_default_conf'", modules))
        ngx.exit(500)
        return
    end
    
    return mod[param]
end

local function _filter_order(_twaf, phase, modules_order)

    local tb = {}
    
    if modules_order[phase] and #modules_order[phase] ~= 0 then
        tb = modules_order[phase]
    else
        tb = modules_order.access
    end
    
    for i = #tb, 1, -1 do
        local mod = _twaf.modfactory[tb[i]]
        if mod and mod[phase] and mod[phase](mod, _twaf) == ngx_DONE then
            break
        end
    end
end

local function _add_resp_headers(_twaf, ctx)
    local cf = _twaf:get_config_param("twaf_add_resp_header") or {}
    ngx_header["X-Tt-Request-Id"] = _twaf:get_vars("UNIQUE_ID", ctx.req)
    
    for k, v in pairs(ctx.add_resp_headers or {}) do
        cf[k] = v
    end
    
    for k, v in pairs(cf) do
        if v ~= "nil" then
            ngx_header[k] = twaf_func:parse_dynamic_value(v, ctx.req)
        end
    end
end

local function _get_response_body(ctx)

    local req = ctx.req
    
    req.RESPONSE_BODY   = ngx.arg[1]
end

function _M.init(self)
    local default_init_register = self:get_default_config_param("init_register")
    _M:register_modules(default_init_register)

    local mod = self.modfactory
    local cf, res, err, policy

    -- access_rule
    cf  = self.config.twaf_access_rule
    res, err = self.modfactory.twaf_access_rule.init(self, cf)
    assert(res, twaf_func:table_to_string(err))

    -- rules -> twaf_conf.lua

    -- twaf_default_conf
    policy = self.config.twaf_default_conf or {}
    for pn, pv in pairs(policy) do
        if mod[pn] and mod[pn]["init"] then
            res, err = mod[pn]["init"](self, pv, policy)
            assert(res, twaf_func:table_to_string(err))
        end
    end

    -- policy
    cf = self.config.twaf_policy.policy_uuids or {}
    for uuid, _ in pairs(cf) do
        policy = self.config.twaf_policy[uuid] or {}
        for pn, pv in pairs(policy) do
            if mod[pn] and mod[pn]["init"] then
                res, err = mod[pn]["init"](self, pv, policy)
                assert(res, twaf_func:table_to_string(err))
            end
        end
    end
end

function _M.run(self, _twaf)

    local res
    local phase         =  ngx_get_phase()
    local modules_order = _twaf:get_config_param("modules_order") or {}
    local ctx           = _twaf:ctx()
    
    ctx.phase       =  phase
    ctx.req.phase_n =  twaf_request.phase[phase]
    
    if not modules_order.access or #modules_order.access == 0 then
        return
    end
    
    if phase == "init_worker" then
        
        for _, modules_name in ipairs(modules_order.access) do
            local mod = _twaf.modfactory[modules_name]
            if mod and mod[phase] then
                mod[phase](_twaf)
            end
        end
    
    elseif phase == "rewrite" then
    
        -- config synchronization
        twaf_func:syn_config(_twaf)
        
        _twaf:get_vars("HTTP_VERSION", ctx.req)
        _twaf:get_vars("RAW_HEADER", ctx.req)
        
        _twaf.modfactory.twaf_access_rule:handler(_twaf)
        
    elseif phase == "access" then
        
        for _, modules_name in ipairs(modules_order.access) do
            local mod = _twaf.modfactory[modules_name]
            if mod and mod["handler"] and mod:handler(_twaf) == ngx_DONE then
                break
            end
        end
        
    elseif phase == "header_filter" then
    
        _filter_order(_twaf, phase, modules_order)
        
        _add_resp_headers(_twaf, ctx)
        
    elseif phase == "body_filter" then
    
       _get_response_body(ctx)
    
        if ctx.reset_connection == true then return ngx_ERROR end
        
         twaf_opts:opts(_twaf, ctx, nil, nil, "respbody_replacement", {})
         
        _filter_order(_twaf, phase, modules_order)
        
        if ctx.reset_connection == true then return ngx_ERROR end
    
    elseif phase == "log" then
    
        -- reqstat
        twaf_reqstat:reqstat_log_handler(_twaf)
        
        -- log
        _twaf.modfactory.twaf_log:log(_twaf)
        
        -- print _G
        twaf_func:print_G(_twaf)
        
        -- print ctx
        twaf_func:print_ctx(_twaf)
        
        -- collect
        -- collectgarbage("collect")
        
    else -- "balancer" "ssl_cert"
    
        local mod = _twaf.modfactory["twaf_"..phase]
        if mod and mod[phase] then
            mod[phase](mod, _twaf)
        end
        
    end
end

return _M
